fft/src/KokkosFFT_ROCM_types.hpp

```diff
@@ -17,27 +26,242 @@ static_assert(sizeof(std::complex<double>) == sizeof(Kokkos::complex<double>));
 constexpr FFTDirectionType ROCFFT_FORWARD  = 1;
 constexpr FFTDirectionType ROCFFT_BACKWARD = -1;
 
+#if !defined(ENABLE_HOST_AND_DEVICE)
 enum class FFTWTransformType { R2C, D2Z, C2R, Z2D, C2C, Z2Z };
+#endif
 
 template <typename ExecutionSpace>
 using TransformType = FFTWTransformType;
 
+/// \brief A class that wraps rocfft_plan_description for RAII
+struct ScopedRocfftPlanDescription {
+ private:
+  rocfft_plan_description m_description;
+
+ public:
+  ScopedRocfftPlanDescription() {
+    rocfft_status status = rocfft_plan_description_create(&m_description);
+    KOKKOSFFT_THROW_IF(status != rocfft_status_success,
+                       "rocfft_plan_description_create failed");
+  }
+  ~ScopedRocfftPlanDescription() noexcept {
+    rocfft_status status = rocfft_plan_description_destroy(m_description);
+    if (status != rocfft_status_success)
+      Kokkos::abort("rocfft_plan_description_destroy failed");
+  }
+
+  rocfft_plan_description description() const noexcept { return m_description; }
+};
+
+/// \brief A class that wraps rocfft_execution_info for RAII
+template <typename FloatingPointType>
+struct ScopedRocfftExecutionInfo {
+ private:
+  using BufferViewType =
+      Kokkos::View<Kokkos::complex<FloatingPointType> *, Kokkos::HIP>;
+  rocfft_execution_info m_execution_info;
+
+  //! Internal work buffer
+  BufferViewType m_buffer;
+
+ public:
+  ScopedRocfftExecutionInfo() {
+    // Prepare workbuffer and set execution information
+    rocfft_status status = rocfft_execution_info_create(&m_execution_info);
+    KOKKOSFFT_THROW_IF(status != rocfft_status_success,
+                       "rocfft_execution_info_create failed");
+  }
+  ~ScopedRocfftExecutionInfo() noexcept {
+    rocfft_status status = rocfft_execution_info_destroy(m_execution_info);
+    if (status != rocfft_status_success)
+      Kokkos::abort("rocfft_execution_info_destroy failed");
+  }
+
+  rocfft_execution_info execution_info() const noexcept {
+    return m_execution_info;
+  }
+
+  void setup(const Kokkos::HIP &exec_space, std::size_t workbuffersize) {
+    // set stream
+    // NOTE: The stream must be of type hipStream_t.
+    // It is an error to pass the address of a hipStream_t object.
+    hipStream_t stream = exec_space.hip_stream();
+    rocfft_status status =
+        rocfft_execution_info_set_stream(m_execution_info, stream);
+    KOKKOSFFT_THROW_IF(status != rocfft_status_success,
+                       "rocfft_execution_info_set_stream failed");
+
+    // Set work buffer
+    if (workbuffersize > 0) {
+      m_buffer = BufferViewType("workbuffer", workbuffersize);
+      status   = rocfft_execution_info_set_work_buffer(
+          m_execution_info, (void *)m_buffer.data(), workbuffersize);
+      KOKKOSFFT_THROW_IF(status != rocfft_status_success,
+                         "rocfft_execution_info_set_work_buffer failed");
+    }
+  }
+};
+
+/// \brief A class that wraps rocfft for RAII
+template <typename T>
```

Comment:
I am not convinced this class needs to be template but this can be discussed after #218

Reply:
Sure, I will do that after this